Add migration steps for `ConfigLoader` to `OmegaConfigLoader` (#2887)
kedro-org/kedro: docs/source/configuration/config_loader_migration.md

# Migration guide for config loaders
The `ConfigLoader` and `TemplatedConfigLoader` classes are now deprecated and scheduled for removal in Kedro `0.19.0`. To ensure smooth transitions for users, we strongly recommend adopting the [`OmegaConfigLoader`](/kedro.config.OmegaConfigLoader) promptly.
This migration guide outlines the primary distinctions between the old loaders and the `OmegaConfigLoader`, providing step-by-step instructions on updating your code base to utilise the new class effectively.

> 

Reviewer: Since `OmegaConfigLoader` is not the default right now, maybe a quick mention here on how to use it by changing `settings.py`. It's mentioned [here](https://docs.kedro.org/en/stable/configuration/advanced_configuration.html#omegaconfigloader) but it's in the advanced section so maybe not everyone goes digging there.

Author: Great point!

Author: Mentioned as the second step below now.

# Migration guide for config loaders
The `ConfigLoader` and `TemplatedConfigLoader` classes have been deprecated since Kedro `0.18.12` and will be removed in Kedro `0.19.0`. To ensure a smooth transition, we strongly recommend you adopt the [`OmegaConfigLoader`](/kedro.config.OmegaConfigLoader) as soon as possible.
This migration guide outlines the primary distinctions between the old loaders and the `OmegaConfigLoader`, providing step-by-step instructions on updating your code base to utilise the new class effectively.

## [`ConfigLoader`](/kedro.config.ConfigLoader) to [`OmegaConfigLoader`](/kedro.config.OmegaConfigLoader)

### 1. Install the Required Library
The [`OmegaConfigLoader`](advanced_configuration.md#omegaconfigloader) was introduced in Kedro `0.18.5` and is based on [OmegaConf](https://omegaconf.readthedocs.io/). In order to use it you need to ensure you have both a version of Kedro of `0.18.5` or above and `omegaconf` installed.
You can install both using `pip`:

```bash
pip install kedro==0.18.5
```

> This would be the minimum required Kedro version which includes `omegaconf` as dependency.

Reviewer 2: ```suggestion
This would be the minimum required Kedro version which includes `omegaconf` as a dependency.
```

Or you can run:
```bash
pip install -U kedro
```

> This command installs the most recent version of Kedro which also includes `omegaconf` as dependency.

Reviewer 2: ```suggestion
This command installs the most recent version of Kedro which also includes `omegaconf` as a dependency.
```

### 2. Use the `OmegaConfigLoader`
To use `OmegaConfigLoader` in your project, set the `CONFIG_LOADER_CLASS` constant in your [`src/<package_name>/settings.py`](../kedro_project_setup/settings.md):

```diff
+ from kedro.config import OmegaConfigLoader  # new import

+ CONFIG_LOADER_CLASS = OmegaConfigLoader
```

### 3. Import Statements
Replace the import statement for `ConfigLoader` with the one for `OmegaConfigLoader`:

```diff
- from kedro.config import ConfigLoader

+ from kedro.config import OmegaConfigLoader
```

### 4. File Format Support

> `OmegaConfigLoader` supports only `yaml` and `json` file formats. Make sure that all your configuration files are in one of these formats. If you were using other formats with `ConfigLoader`, convert them to `yaml` or `json`.

Reviewer 2: ```suggestion
`OmegaConfigLoader` supports only `yaml` and `json` file formats. Make sure that all your configuration files are in one of these formats. If you previously used other formats with `ConfigLoader`, convert them to `yaml` or `json`.
```

### 5. Load Configuration
The method to load the configuration using `OmegaConfigLoader` differs slightly from that used by `ConfigLoader`, which allowed users to access configuration through the `.get()` method and required patterns as argument.

> When you migrate to use `OmegaConfigLoader`it  requires you to fetch configuration through a configuration key that points to [configuration patterns specified in the loader class](configuration_basics.md#configuration-patterns) or [provided in the `CONFIG_LOADER_ARGS`](advanced_configuration.md#how-to-change-which-configuration-files-are-loaded) in `settings.py`.

Reviewer 2: ```suggestion
When you migrate to use `OmegaConfigLoader` it  requires you to fetch configuration through a configuration key that points to [configuration patterns specified in the loader class](configuration_basics.md#configuration-patterns) or [provided in the `CONFIG_LOADER_ARGS`](advanced_configuration.md#how-to-change-which-configuration-files-are-loaded) in `settings.py`.
```

```diff
- conf_path = str(project_path / settings.CONF_SOURCE)
- conf_loader = ConfigLoader(conf_source=conf_path, env="local")
- catalog = conf_loader.get("catalog*")

+ conf_path = str(project_path / settings.CONF_SOURCE)
+ config_loader = OmegaConfigLoader(conf_source=conf_path, env="local")
+ catalog = config_loader["catalog"]
```

In this example, `"catalog"` is the key to the default catalog patterns specified in the `OmegaConfigLoader` class.

> ### 6. Exception Handling

Reviewer 2: Am I correct in thinking this guide is mostly flagging differences rather than a full set of migration steps. So, in this section, you're showing where errors could be different. Are there things that are the same that you need to call out, as in "For error and exception handling, most errors are the same. Those you need to be aware of that are different between the original `ConfigLoader` and `OmegaConfigLoader` are as follows:"

Or is everything different, and set out below ? I'm just thinking about setting user expectations I guess.

Author: That's a good point. I have indeed mostly focussed on differences, because all the other behaviour is still the same.  This section about errors is quite specific and most users probably won't care, but I added it in case users are catching a specific type of error and need to now change that. I'll add the sentence in mentioning most errors are still the same.

This migration guide outlines the primary distinctions between the old loaders and the `OmegaConfigLoader`, providing step-by-step instructions on updating your code base to utilise the new class effectively.

## [`ConfigLoader`](/kedro.config.ConfigLoader) to [`OmegaConfigLoader`](/kedro.config.OmegaConfigLoader)

### 1. Install the Required Library
The [`OmegaConfigLoader`](advanced_configuration.md#omegaconfigloader) was introduced in Kedro `0.18.5` and is based on [OmegaConf](https://omegaconf.readthedocs.io/). In order to use it you need to ensure you have both a version of Kedro of `0.18.5` or above and `omegaconf` installed.
You can install both using `pip`:

```bash
pip install kedro==0.18.5
```
This would be the minimum required Kedro version which includes `omegaconf` as dependency.
Or you can run:
```bash
pip install -U kedro
```

This command installs the most recent version of Kedro which also includes `omegaconf` as dependency.

### 2. Use the `OmegaConfigLoader`
To use `OmegaConfigLoader` in your project, set the `CONFIG_LOADER_CLASS` constant in your [`src/<package_name>/settings.py`](../kedro_project_setup/settings.md):

> ```python
> from kedro.config import OmegaConfigLoader  # new import
> 
> CONFIG_LOADER_CLASS = OmegaConfigLoader

Reviewer 3: ```suggestion
```diff
+ from kedro.config import OmegaConfigLoader

+ CONFIG_LOADER_CLASS = OmegaConfigLoader
```
What about just show the diff? Not sure if we can keep the syntax highlight with diff

Author: Ahh I didn't know that existed! Let me try 🙂

```

### 3. Import Statements
Replace the import statement for `ConfigLoader` with the one for `OmegaConfigLoader`:

```python
# Before:
from kedro.config import ConfigLoader

# After:

> from kedro.config import OmegaConfigLoader

Reviewer 3: Same can we use diff here to indicate the change?

Author: Done!

```

### 4. File Format Support
`OmegaConfigLoader` supports only `yaml` and `json` file formats. Make sure that all your configuration files are in one of these formats. If you were using other formats with `ConfigLoader`, convert them to `yaml` or `json`.

### 5. Load Configuration
The method to load the configuration using `OmegaConfigLoader` differs slightly from that used by `ConfigLoader`, which allowed users to access configuration through the `.get()` method and required patterns as argument.
When you migrate to use `OmegaConfigLoader`it  requires you to fetch configuration through a configuration key that points to [configuration patterns specified in the loader class](configuration_basics.md#configuration-patterns) or [provided in the `CONFIG_LOADER_ARGS`](advanced_configuration.md#how-to-change-which-configuration-files-are-loaded) in `settings.py`.

```python
# Before:

> conf_path = str(project_path / settings.CONF_SOURCE)

Reviewer 3: str is not necessary?

Author: I think it might be because `settings.CONF_SOURCE` is a string so you probably get an error if you try concatenating a `Path` and string with `/`

conf_loader = ConfigLoader(conf_source=conf_path, env="local")
catalog = conf_loader.get("catalog*")

# After:
conf_path = str(project_path / settings.CONF_SOURCE)
config_loader = OmegaConfigLoader(conf_source=conf_path, env="local")

> catalog = config_loader["catalog"]

Reviewer 3: Isn't the pattern and dictionary interface available for all config loader? Maybe I remember the wrong thing.

Author: Yes it is. I put this in because currently both `.get("catalog*")` and `[catalog]` works for `ConfigLoader` but for `OmegaConfigLoader` only `[catalog]` works. I don't know if users have already switched to the newer dictionary interface.
